tests/numpy/test_creation_manipulation.py

```diff
@@ -78,6 +79,37 @@ def test_diag(nps_app_inst):
     np_arr = np.diag(np_arr)
     assert np.allclose(ba.get(), np_arr)
 
+def test_trace(nps_app_inst):
+    import nums.numpy as nps
+
+    assert nps_app_inst is not None
+
+    a: BlockArray = nps.array([1.0, 2.0, 3.0, 4.0])
+
+    # Test that trace doesn't work with 1-dimensional vectors
+    with pytest.raises(ValueError):
+        nps.trace(a)
+        
+    # Create Diagonal Matrices
+    a_diag = nps.diag(a) # Nums Version
+    a_diag_np = np.diag(a.get()) # Numpy version
+
+    # Apply on Trace on each of the Diagonal matrices
+    a_diag_trace = nps.trace(a_diag).get()
+    a_diag_np_trace = np.trace(a_diag_np)
+
+    assert np.allclose(a_diag_trace, a_diag_np_trace)
+
+    # Use pre-created diagonal matrices
+    b: BlockArray = nps.array([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0], [7.0, 8.0, 9.0]])
```

Comment:
Can you test non-square matrices (short-wide and tall-skinny)?

Reply:
The behavior of NumPy and NumS is different for diagonal matrices. Doing the following 
`>>> a = np.array([[1, 2], [3,4], [5,6]])
>>> np.diag(a)
array([1, 4])
`
But doing the equivalent in NumS, produces an assertion error. 
`>>> a: BlockArray = nps.array([[1, 2], [3, 4], [5, 6]])
>>> nps.diag(a)
AssertionError`

This is triggered by the line `assert X.shape[0] == X.shape[1]` from Line 423 in core/array/application.py, enforcing that X needs to be square.

Reply:
Indeed, let's discuss offline further to decide whether we want to work on diag to support these cases. I can help with the diag imp if needed.